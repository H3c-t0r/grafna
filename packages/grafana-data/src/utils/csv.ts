// Libraries
import Papa, { ParseResult, ParseConfig, Parser } from 'papaparse';
import defaults from 'lodash/defaults';
import isNumber from 'lodash/isNumber';

// Types
import { DataFrame, Field, FieldType, FieldSchema } from '../types';
import { guessFieldTypeFromValue } from './processDataFrame';
import { ArrayVector } from './vector';

export enum CSVHeaderStyle {
  full,
  name,
  none,
}

// Subset of all parse options
export interface CSVConfig {
  delimiter?: string; // default: ","
  newline?: string; // default: "\r\n"
  quoteChar?: string; // default: '"'
  encoding?: string; // default: "",
  headerStyle?: CSVHeaderStyle;
}

export interface CSVParseCallbacks {
  /**
   * Get a callback before any rows are processed
   * This can return a modified table to force any
   * Column configurations
   */
  onHeader: (table: DataFrame) => void;

  // Called after each row is read and
  onRow: (row: any[]) => void;
}

export interface CSVOptions {
  config?: CSVConfig;
  callback?: CSVParseCallbacks;
}

export function readCSV(csv: string, options?: CSVOptions): DataFrame[] {
  return new CSVReader(options).readCSV(csv);
}

enum ParseState {
  Starting,
  InHeader,
  ReadingRows,
}

type FieldParser = (value: string) => any;

type AnyField = Field<any, ArrayVector<any>>;

export class CSVReader {
  config: CSVConfig;
  callback?: CSVParseCallbacks;

  field: FieldParser[];
  state: ParseState;
  data: DataFrame[];
  fields: AnyField[];

  constructor(options?: CSVOptions) {
    if (!options) {
      options = {};
    }
    this.config = options.config || {};
    this.callback = options.callback;

    this.field = [];
    this.state = ParseState.Starting;
    this.fields = [];
    this.data = [];
  }

  // PapaParse callback on each line
  private step = (results: ParseResult, parser: Parser): void => {
    for (let i = 0; i < results.data.length; i++) {
      const line: string[] = results.data[i];
      if (line.length < 1) {
        continue;
      }
      const first = line[0]; // null or value, papaparse does not return ''
      if (first) {
        // Comment or header queue
        if (first.startsWith('#')) {
          // Look for special header column
          // #{columkey}#a,b,c
          const idx = first.indexOf('#', 2);
          if (idx > 0) {
            const k = first.substr(1, idx - 1);
            const isName = 'name' === k;

            // Simple object used to check if headers match
            const headerKeys: FieldSchema = {
              type: FieldType.number,
              unit: '#',
              dateSourceFormat: '#',
            };

            // Check if it is a known/supported column
            if (isName || headerKeys.hasOwnProperty(k)) {
              // Starting a new table after reading rows
              if (this.state === ParseState.ReadingRows) {
                this.data.push({
                  fields: this.fields,
                });
                this.fields = [];
              }

              // Fix any column sizing issues
              padColumnWidth(this.fields, line.length);

              const v = first.substr(idx + 1);
              if (isName) {
                this.fields[0].name = v;
                for (let j = 1; j < this.fields.length; j++) {
                  this.fields[j].name = line[j];
                }
              } else {
                for (let j = 0; j < this.fields.length; j++) {
                  const schema = this.fields[j].schema as any; // any lets name lookup
                  schema[k] = j === 0 ? v : line[j];
                }
              }

              this.state = ParseState.InHeader;
              continue;
            }
          } else if (this.state === ParseState.Starting) {
            this.fields = makeFieldsFor(line);
            this.state = ParseState.InHeader;
            continue;
          }
          // Ignore comment lines
          continue;
        }

        if (this.state === ParseState.Starting) {
          const type = guessFieldTypeFromValue(first);
          if (type === FieldType.string) {
            this.fields = makeFieldsFor(line);
            this.state = ParseState.InHeader;
            continue;
          }
          this.fields = makeFieldsFor(new Array(line.length));
          this.fields[0].schema.type = type;
          this.state = ParseState.InHeader; // fall through to read rows
        }
      }

      if (this.state === ParseState.InHeader) {
        padColumnWidth(this.fields, line.length);
        this.state = ParseState.ReadingRows;
      }

      if (this.state === ParseState.ReadingRows) {
        // Make sure colum structure is valid
        if (line.length > this.fields.length) {
          padColumnWidth(this.fields, line.length);
          if (this.callback) {
            this.callback.onHeader({ fields: this.fields });
          }
        }

        const row: any[] = [];
        for (let j = 0; j < line.length; j++) {
          let v: any = line[j];
          if (v) {
            if (!this.field[j]) {
              this.field[j] = makeFieldParser(v, this.fields[j]);
            }
            v = this.field[j](v);
          } else {
            v = null;
          }
          this.fields[j].values.buffer.push(v);
          if (this.callback) {
            row.push(v);
          }
        }

        if (this.callback) {
          // // Send the header after we guess the type
          // if (this.series.rows.length === 0) {
          //   this.callback.onHeader(this.series);
          //   this.series.rows.push(row); // Only add the first row
          // }
          this.callback.onRow(row);
        }
      }
    }
  };

  readCSV(text: string): DataFrame[] {
    this.data = [];

    const papacfg = {
      ...this.config,
      dynamicTyping: false,
      skipEmptyLines: true,
      comments: false, // Keep comment lines
      step: this.step,
    } as ParseConfig;

    Papa.parse(text, papacfg);
    return this.data;
  }
}

function makeFieldParser(value: string, field: Field): FieldParser {
  let schema = field.schema;
  if (!schema || !schema.type) {
    if (!schema) {
      field.schema = schema = {};
    }
    if (field.name === 'time' || field.name === 'Time') {
      schema.type = FieldType.time;
    } else {
      schema.type = guessFieldTypeFromValue(value);
    }
  }

  if (schema.type === FieldType.number) {
    return (value: string) => {
      return parseFloat(value);
    };
  }

  // Will convert anything that starts with "T" to true
  if (schema.type === FieldType.boolean) {
    return (value: string) => {
      return !(value[0] === 'F' || value[0] === 'f' || value[0] === '0');
    };
  }

  // Just pass the string back
  return (value: string) => value;
}

/**
 * Creates a field object for each string in the list
 */
function makeFieldsFor(line: string[]): AnyField[] {
  const fields: AnyField[] = [];
  for (let i = 0; i < line.length; i++) {
    const v = line[i] ? line[i] : 'Column ' + (i + 1);
    fields.push({ name: v, schema: {}, values: new ArrayVector([]) });
  }
  return fields;
}

type FieldWriter = (value: any) => string;

function writeValue(value: any, config: CSVConfig): string {
  const str = value.toString();
  if (str.includes('"')) {
    // Escape the double quote characters
    return config.quoteChar + str.replace(/"/gi, '""') + config.quoteChar;
  }
  if (str.includes('\n') || str.includes(config.delimiter)) {
    return config.quoteChar + str + config.quoteChar;
  }
  return str;
}

function makeFieldWriter(field: Field, config: CSVConfig): FieldWriter {
  const { schema } = field;
  if (schema.type) {
    if (schema.type === FieldType.boolean) {
      return (value: any) => {
        return value ? 'true' : 'false';
      };
    }

    if (schema.type === FieldType.number) {
      return (value: any) => {
        if (isNumber(value)) {
          return value.toString();
        }
        return writeValue(value, config);
      };
    }
  }

  return (value: any) => writeValue(value, config);
}

function getHeaderLine(key: string, fields: Field[], config: CSVConfig): string {
  const isName = 'name' === key;

  for (const f of fields) {
    const schema = f.schema;
    if (isName || schema.hasOwnProperty(key)) {
      let line = '#' + key + '#';
      for (let i = 0; i < fields.length; i++) {
        if (i > 0) {
          line = line + config.delimiter;
        }

        const v = isName ? f.name : (fields[i].schema as any)[key];
        if (v) {
          line = line + writeValue(v, config);
        }
      }
      return line + config.newline;
    }
  }
  return '';
}

export function toCSV(data: DataFrame[], config?: CSVConfig): string {
  if (!data) {
    return '';
  }

  let csv = '';
  config = defaults(config, {
    delimiter: ',',
    newline: '\r\n',
    quoteChar: '"',
    encoding: '',
    headerStyle: CSVHeaderStyle.name,
  });

  for (const series of data) {
    const { fields } = series;
    if (config.headerStyle === CSVHeaderStyle.full) {
      csv =
        csv +
        getHeaderLine('name', fields, config) +
        getHeaderLine('type', fields, config) +
        getHeaderLine('unit', fields, config) +
        getHeaderLine('dateFormat', fields, config);
    } else if (config.headerStyle === CSVHeaderStyle.name) {
      for (let i = 0; i < fields.length; i++) {
        if (i > 0) {
          csv += config.delimiter;
        }
        csv += fields[i].name;
      }
      csv += config.newline;
    }
    const length = fields[0].values.length;
    if (length > 0) {
      const writers = fields.map(field => makeFieldWriter(field, config!));
      for (let i = 0; i < length; i++) {
        for (let j = 0; j < fields.length; j++) {
          if (j > 0) {
            csv = csv + config.delimiter;
          }

          const v = fields[j].values.get(i);
          if (v !== null) {
            csv = csv + writers[j](v);
          }
        }
        csv = csv + config.newline;
      }
    }
    csv = csv + config.newline;
  }

  return csv;
}

/**
 * Make sure a field exists for the entire width and fill any extras with null
 */
function padColumnWidth(fields: AnyField[], width: number) {
  if (fields.length === width) {
    return; // no changes
  }

  // Find the maximum length of all the fields
  const length = fields.reduce((prev, field) => {
    return Math.max(prev, field.values.length);
  }, 0);

  // Add extra columns
  for (let i = fields.length; i < width; i++) {
    fields.push({
      name: `Field ${i + 1}`,
      schema: {},
      values: new ArrayVector([]),
    });
  }

  // Fill the column values
  for (const field of fields) {
    if (field.values.length < length) {
      for (let i = field.values.length; i < length; i++) {
        field.values.buffer.push(null);
      }
    }
  }
}
