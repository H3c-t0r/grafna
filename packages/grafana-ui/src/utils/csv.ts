// Libraries
import Papa, { ParseResult, ParseConfig, Parser } from 'papaparse';

// Types
import { SeriesData, Field, FieldType } from '../types/index';
import { guessFieldTypeFromValue } from './processSeriesData';

// Subset of all parse options
export interface CSVParseConfig {
  delimiter?: string; // default: ","
  newline?: string; // default: "\r\n"
  quoteChar?: string; // default: '"'
  encoding?: string; // default: ""
}

export interface CSVParseCallbacks {
  /**
   * Get a callback before any rows are processed
   * This can return a modified table to force any
   * Column configurations
   */
  onHeader: (table: SeriesData) => void;

  // Called after each row is read and
  onRow: (row: any[]) => void;
}

export interface CSVOptions {
  config?: CSVParseConfig;
  callback?: CSVParseCallbacks;
}

export function readCSV(csv: string, options?: CSVOptions): Promise<SeriesData[]> {
  // Wraps the string in a ReadableStreamReader
  return readCSVFromStream(
    {
      cancel: () => {
        return Promise.reject('unsupported');
      },
      read: () => {
        return Promise.resolve({ done: true, value: csv });
      },
      releaseLock: () => {},
    },
    options
  );
}

enum ParseState {
  Starting,
  InHeader,
  ReadingRows,
}

type FieldParser = (value: string) => any;

export function readCSVFromStream(reader: ReadableStreamReader<string>, options?: CSVOptions): Promise<SeriesData[]> {
  return new Promise((resolve, reject) => {
    const config = options ? options.config : {};
    const callback = options ? options.callback : null;

    const column: FieldParser[] = [];
    let state = ParseState.Starting;
    let table: SeriesData = {
      fields: [],
      rows: [],
    };
    const tables: SeriesData[] = [table];

    const step = (results: ParseResult, parser: Parser): void => {
      for (let i = 0; i < results.data.length; i++) {
        const line: string[] = results.data[i];
        if (line.length > 0) {
          const first = line[0]; // null or value, papaparse does not return ''
          if (first) {
            // Comment or header queue
            if (first.startsWith('#')) {
              // Look for special header column
              // #{columkey}#a,b,c
              const idx = first.indexOf('#', 2);
              if (idx > 0) {
                const k = first.substr(1, idx - 1);

                // Simple object used to check if headers match
                const headerKeys: Field = {
                  name: '#',
                  type: FieldType.number,
                  unit: '#',
                  dateFormat: '#',
                };

                // Check if it is a known/supported column
                if (headerKeys.hasOwnProperty(k)) {
                  // Starting a new table after reading rows
                  if (state === ParseState.ReadingRows) {
                    table = {
                      fields: [],
                      rows: [],
                    };
                    tables.push(table);
                  }

                  padColumnWidth(table.fields, line.length);
                  const fields: any[] = table.fields; // cast to any so we can lookup by key
                  const v = first.substr(idx + 1);
                  fields[0][k] = v;
                  for (let j = 1; j < fields.length; j++) {
                    fields[j][k] = line[j];
                  }
                  state = ParseState.InHeader;
                  continue;
                }
              } else if (state === ParseState.Starting) {
                table.fields = makeFieldsFor(line);
                state = ParseState.InHeader;
                continue;
              }
              // Ignore comment lines
              continue;
            }

            if (state === ParseState.Starting) {
              const type = guessFieldTypeFromValue(first);
              if (type === FieldType.string) {
                table.fields = makeFieldsFor(line);
                state = ParseState.InHeader;
                continue;
              }
              table.fields = makeFieldsFor(new Array(line.length));
              table.fields[0].type = type;
              state = ParseState.InHeader; // fall through to read rows
            }
          }

          if (state === ParseState.InHeader) {
            padColumnWidth(table.fields, line.length);
            state = ParseState.ReadingRows;
          }

          if (state === ParseState.ReadingRows) {
            // Make sure colum structure is valid
            if (line.length > table.fields.length) {
              padColumnWidth(table.fields, line.length);
              if (callback) {
                callback.onHeader(table);
              } else {
                // Expand all rows with nulls
                for (let x = 0; x < table.rows.length; x++) {
                  const row = table.rows[x];
                  while (row.length < line.length) {
                    row.push(null);
                  }
                }
              }
            }

            const row: any[] = [];
            for (let j = 0; j < line.length; j++) {
              const v = line[j];
              if (v) {
                if (!column[j]) {
                  column[j] = makeFieldParser(v, table.fields[j]);
                }
                row.push(column[j](v));
              } else {
                row.push(null);
              }
            }

            if (callback) {
              // Send the header after we guess the type
              if (table.rows.length === 0) {
                callback.onHeader(table);
                table.rows.push(row); // Only add the first row
              }
              callback.onRow(row);
            } else {
              table.rows.push(row);
            }
          }
        }
      }
    };

    const papacfg = {
      ...config,
      dynamicTyping: false,
      skipEmptyLines: true,
      comments: false, // Keep comment lines
      step,
    } as ParseConfig;

    const processText = (value: ReadableStreamReadResult<string>): any => {
      if (value.value) {
        Papa.parse(value.value, papacfg);
      }
      if (value.done) {
        resolve(tables);
        return;
      }
      return reader.read().then(processText);
    };
    reader.read().then(processText);
  });
}

function makeFieldParser(value: string, column: Field): FieldParser {
  if (!column.type) {
    column.type = guessFieldTypeFromValue(value);
  }

  if (column.type === FieldType.number) {
    return (value: string) => {
      return parseFloat(value);
    };
  }

  // Will convert anything that starts with "T" to true
  if (column.type === FieldType.boolean) {
    return (value: string) => {
      return !(value[0] === 'F' || value[0] === 'f' || value[0] === '0');
    };
  }

  // Just pass the string back
  return (value: string) => value;
}

/**
 * Creates a column object for each string in the list
 */
function makeFieldsFor(line: string[]): Field[] {
  const fields: Field[] = [];
  for (let i = 0; i < line.length; i++) {
    const v = line[i] ? line[i] : 'Column ' + (i + 1);
    fields.push({ name: v });
  }
  return fields;
}

/**
 * Makes sure the colum has valid entries up the the width
 */
function padColumnWidth(fields: Field[], width: number) {
  if (fields.length < width) {
    for (let i = fields.length; i < width; i++) {
      fields.push({
        name: 'Field ' + (i + 1),
      });
    }
  }
}
