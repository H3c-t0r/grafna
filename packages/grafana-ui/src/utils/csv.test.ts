import { readCSV, readCSVFromStream } from './csv';

const fs = require('fs');
const Readable = require('stream').Readable;

describe('read csv', () => {
  it('should get X and y', () => {
    const text = ',1\n2,3,4\n5,6\n,,,7';
    return readCSV(text).then(tables => {
      expect(tables.length).toBe(1);

      const table = tables[0];
      expect(table.columns.length).toBe(4);
      expect(table.rows.length).toBe(3);

      // Make sure everythign it padded properly
      for (const row of table.rows) {
        expect(row.length).toBe(table.columns.length);
      }

      expect(tables[0]).toMatchSnapshot();
    });
  });

  it('should read csv from local file system', () => {
    const path = __dirname + '/testdata/simple.csv';
    console.log('PATH', path);
    expect(fs.existsSync(path)).toBeTruthy();

    const stream = fs.createReadStream(path, 'utf8');
    return readCSVFromStream(stream).then(tables => {
      //expect(tables.length).toBe(1);
      expect(tables[0]).toMatchSnapshot();
    });
  });
});
