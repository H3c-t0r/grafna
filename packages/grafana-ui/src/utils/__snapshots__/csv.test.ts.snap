// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read csv should get X and y 1`] = `
Object {
  "columns": Array [
    Object {
      "text": "Column 1",
      "type": "number",
    },
    Object {
      "text": "Column 2",
      "type": "number",
    },
    Object {
      "text": "Column 3",
      "type": "number",
    },
    Object {
      "text": "Column 4",
      "type": "number",
    },
  ],
  "rows": Array [
    Array [
      2,
      3,
      4,
      null,
    ],
    Array [
      5,
      6,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
      7,
    ],
  ],
}
`;

exports[`read csv should read csv from local file system 1`] = `
Object {
  "columns": Array [],
  "rows": Array [],
}
`;
